### DISCUSSION TITLE
Use Stream with external texture on Android API 24

### DISCUSSION THREAD

**Discussion by kpeeters** [google/filament#3878] [General]
I am trying to use a `filament::Stream` connected to an external texture using `.stream` on Android API 24. Am using filament built from source using the standard `build.sh -p android -i release|debug`. On API 30 this all works fine, but on 24 the debug build bails out with:

    04-30 19:41:51.685 22087 22116 E Filament: in void utils::panic(const char *, const char *, int, const char *):32
    04-30 19:41:51.685 22087 22116 E Filament: in file ../../libs/utils/src/debug.cpp
    04-30 19:41:51.685 22087 22116 E Filament: reason: ../../filament/backend/src/android/VirtualMachineEnv.h:61: failed assertion `mVirtualMachine'
    04-30 19:41:51.685 22087 22116 E Filament: 

Looking around a bit in `VirtualMachineEnv.h` and friends shows that there is special code for API < 26, but that only gets included at compile time (at which time it probably looks at `targetSdkVersion`, which is probably >=26?).

What is the recommended way to get this working on API 24, preferably with a build that also still works on API >=26 without sacrificing other things? I don't mind building filament myself (am already doing that anyway). Or is this a bug?

**Comment by romainguy** [comment]
This looks like an initialization problem. Are you properly calling `Filament.init()`? It looks like `VirtualMachineEnv::JNI_OnLoad()` wasn't invoked. The compile-time check for `AHardwareBuffer` shouldn't matter since we do the resolution at runtime otherwise.

**Reply by kpeeters** [reply]
Thanks, that did the trick! I do all initialisation on the native side so I do not link to any of your JNI bridging code. And thus I missed that you have your JNI_OnLoad with that non-trivial `::filament::OnLoad` call in it. Adding that to my `JNI_OnLoad` makes it all work.

Can I donate a few beers?

**Reply by romainguy** [reply]
You may run into similar things if you use our other libraries (gltfio, etc.).

As for beers, we're just happy to help :)
